cluster-autoscaler/cloudprovider/gce/cache.go

```diff
@@ -165,71 +160,110 @@ func (gc *GceCache) updateMigBasename(ref GceRef, basename string) {
 // Attempts to regenerate cache if there is a Mig with matching prefix in migs list.
 // TODO(aleksandra-malinowska): reconsider failing when there's a Mig with
 // matching prefix, but instance doesn't belong to it.
-func (gc *GceCache) GetMigForInstance(instance *GceRef) (Mig, error) {
+func (gc *GceCache) GetMigForInstance(instanceRef GceRef) (Mig, error) {
 	gc.cacheMutex.Lock()
 	defer gc.cacheMutex.Unlock()
 
-	if mig, found := gc.instancesCache[*instance]; found {
-		return mig, nil
+	if migRef, found := gc.instanceRefToMigRef[instanceRef]; found {
+		mig, found := gc.migs[migRef]
+		if !found {
+			// It is fine that mig is not found in gc.migs map even if we have entry in gc.instanceRefToMigRef
+			// The mig could have bin removed rom gc.migs if we (for example) disabled autoscaling for that or the node pool
+			// was just deleted. Yet we are not removing entries from gc.instanceRefToMigRef in such case due to performance reasons.
+			return nil, nil
+		}
+		return mig.Config, nil
 	}
 
 	for _, mig := range gc.GetMigs() {
-		if mig.Config.GceRef().Project == instance.Project &&
-			mig.Config.GceRef().Zone == instance.Zone &&
-			strings.HasPrefix(instance.Name, mig.Basename) {
-			if err := gc.regenerateCache(); err != nil {
-				return nil, fmt.Errorf("error while looking for MIG for instance %+v, error: %v", *instance, err)
+		if mig.Config.GceRef().Project == instanceRef.Project &&
+			mig.Config.GceRef().Zone == instanceRef.Zone &&
+			strings.HasPrefix(instanceRef.Name, mig.Basename) {
+			if err := gc.regenerateInstanceCacheForMigNoLock(mig.Config.GceRef()); err != nil {
+				return nil, fmt.Errorf("error while looking for MIG for instance %+v, error: %v", instanceRef, err)
 			}
-			if mig, found := gc.instancesCache[*instance]; found {
-				return mig, nil
+
+			migRef, found := gc.instanceRefToMigRef[instanceRef]
+			if !found {
+				return nil, fmt.Errorf("instance %+v belongs to unknown mig %+v", instanceRef, migRef)
+			}
+			mig, found := gc.migs[migRef]
```

Comment:
Can we add some method that will retrieve mig under migsMutex lock?

Reply:
Done - seem comments above